File: apps/web-admin/src/app/features/map/utils/map-lightning-effect.ts
/** Destello de relámpago esporádico con rayo ramificado (solo con lluvia intensa). */

interface BoltPoint {
  x: number;
  y: number;
}

export class MapLightningEffect {
  private flashAge = 0;
  private flashDuration = 0;
  private cooldown = 90;
  private enabled = false;
  private rainIntensity = 0;
  /** Rayo principal + ramas, en coords [0..1] del viewport; null = sin rayo. */
  private bolt: BoltPoint[] | null = null;
  private branches: BoltPoint[][] = [];
  private boltSide = 0.5;

  setEnabled(value: boolean): void {
    this.enabled = value;
    if (!value) {
      this.flashAge = 0;
      this.flashDuration = 0;
      this.bolt = null;
      this.branches = [];
    }
  }

  setRainIntensity(value: number): void {
    this.rainIntensity = Math.min(1, Math.max(0, value));
  }

  clear(): void {
    this.flashAge = 0;
    this.flashDuration = 0;
    this.cooldown = 60;
    this.bolt = null;
    this.branches = [];
  }

  /** Genera un rayo quebrado de arriba hacia abajo con 1-2 ramas. */
  private generateBolt(): void {
    this.boltSide = 0.2 + Math.random() * 0.6;
    const segments = 7 + Math.floor(Math.random() * 4);
    const main: BoltPoint[] = [{ x: this.boltSide, y: 0 }];
    let x = this.boltSide;
    const endY = 0.5 + Math.random() * 0.32;
    for (let i = 1; i <= segments; i++) {
      const t = i / segments;
      x += (Math.random() - 0.5) * 0.11;
      x = Math.min(0.95, Math.max(0.05, x));
      main.push({ x, y: t * endY });
    }
    this.bolt = main;

    this.branches = [];
    const branchCount = 1 + Math.floor(Math.random() * 2);
    for (let b = 0; b < branchCount; b++) {
      const startIdx = 2 + Math.floor(Math.random() * (main.length - 3));
      const start = main[startIdx];
      const branch: BoltPoint[] = [{ ...start }];
      let bx = start.x;
      let by = start.y;
      const dir = Math.random() > 0.5 ? 1 : -1;
      const steps = 2 + Math.floor(Math.random() * 3);
      for (let i = 0; i < steps; i++) {
        bx += dir * (0.03 + Math.random() * 0.06);
        by += 0.04 + Math.random() * 0.07;
        bx = Math.min(0.97, Math.max(0.03, bx));
        branch.push({ x: bx, y: by });
      }
      this.branches.push(branch);
    }
  }

  tick(rainActive: boolean, dt = 1): void {
    if (this.flashDuration > 0) {
      this.flashAge += dt;
      if (this.flashAge >= this.flashDuration) {
        this.flashAge = 0;
        this.flashDuration = 0;
        this.bolt = null;
        this.branches = [];
        this.cooldown = 70 + Math.random() * 120;
      }
      return;
    }

    if (!this.enabled || !rainActive || this.rainIntensity < 0.7) return;

    this.cooldown -= dt;
    if (this.cooldown > 0) return;

    if (Math.random() > 0.018 + (this.rainIntensity - 0.7) * 0.06) return;

    this.flashDuration = 4 + Math.random() * 6;
    this.flashAge = 0;
    this.cooldown = 80 + Math.random() * 140;
    // ponytail: ~55% de los destellos muestran rayo visible; el resto es solo fogonazo lejano.
    this.bolt = null;
    this.branches = [];
    if (Math.random() < 0.55) this.generateBolt();
  }

  isFlashing(): boolean {
    return this.flashDuration > 0 && this.flashAge < this.flashDuration;
  }

  /** Dispara un destello con rayo de inmediato (previsualización / pruebas). */
  forceFlash(withBolt = true): void {
    this.flashDuration = 4 + Math.random() * 6;
    this.flashAge = 0;
    if (withBolt) this.generateBolt();
    else { this.bolt = null; this.branches = []; }
  }

  private strokeBolt(
    ctx: CanvasRenderingContext2D,
    pts: BoltPoint[],
    w: number,
    h: number,
    width: number,
    color: string,
  ): void {
    if (pts.length < 2) return;
    ctx.strokeStyle = color;
    ctx.lineWidth = width;
    ctx.lineCap = 'round';
    ctx.lineJoin = 'round';
    ctx.beginPath();
    ctx.moveTo(pts[0].x * w, pts[0].y * h);
    for (let i = 1; i < pts.length; i++) ctx.lineTo(pts[i].x * w, pts[i].y * h);
    ctx.stroke();
  }

  draw(
    ctx: CanvasRenderingContext2D,
    clipPath: Path2D | null,
    viewportW: number,
    viewportH: number,
  ): void {
    if (!this.isFlashing()) return;

    const t = this.flashAge / this.flashDuration;
    const peak = t < 0.25 ? t / 0.25 : 1 - (t - 0.25) / 0.75;
    const alpha = peak * 0.42;

    ctx.save();
    if (clipPath) ctx.clip(clipPath);

    // Fogonazo cálido (cómic): blanco-amarillo
    ctx.fillStyle = `rgba(255, 250, 215, ${alpha * 0.5})`;
    ctx.fillRect(0, 0, viewportW, viewportH);
    ctx.fillStyle = `rgba(255, 235, 150, ${alpha * 0.3})`;
    ctx.fillRect(0, viewportH * 0.06, viewportW, viewportH * 0.35);

    if (this.bolt) {
      const boltFade = Math.max(0, 1 - t * 1.4);
      if (boltFade > 0.02) {
        const scale = Math.min(viewportW, viewportH) / 320 + 0.6;
        // Contorno negro grueso (estilo cómic)
        this.strokeBolt(ctx, this.bolt, viewportW, viewportH, 7 * scale, `rgba(10, 10, 20, ${boltFade})`);
        for (const br of this.branches) this.strokeBolt(ctx, br, viewportW, viewportH, 4.5 * scale, `rgba(10, 10, 20, ${boltFade})`);
        // Relleno amarillo
        this.strokeBolt(ctx, this.bolt, viewportW, viewportH, 4 * scale, `rgba(255, 214, 40, ${boltFade})`);
        for (const br of this.branches) this.strokeBolt(ctx, br, viewportW, viewportH, 2.4 * scale, `rgba(255, 214, 40, ${boltFade})`);
        // Núcleo blanco caliente
        this.strokeBolt(ctx, this.bolt, viewportW, viewportH, 1.6 * scale, `rgba(255, 255, 240, ${boltFade})`);
        for (const br of this.branches) this.strokeBolt(ctx, br, viewportW, viewportH, 1 * scale, `rgba(255, 255, 240, ${boltFade * 0.9})`);
      }
    }

    ctx.restore();
  }
}

File: apps/mobile-android/app/src/main/java/com/univalle/pedrochacolla/ui/dashboard/MapLightningEffect.kt
package com.univalle.pedrochacolla.ui.dashboard

import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Path
import kotlin.random.Random

class MapLightningEffect {
    private data class BoltPoint(val x: Float, val y: Float)

    private var flashAge = 0f
    private var flashDuration = 0f
    private var cooldown = 90f
    private var enabled = false
    private var rainIntensity = 0f
    private var bolt: List<BoltPoint>? = null
    private var branches: List<List<BoltPoint>> = emptyList()
    private val strokePaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
        style = Paint.Style.STROKE
        strokeCap = Paint.Cap.ROUND
        strokeJoin = Paint.Join.ROUND
    }
    private val fillPaint = Paint(Paint.ANTI_ALIAS_FLAG)

    fun setEnabled(value: Boolean) {
        enabled = value
        if (!value) clear()
    }

    fun setRainIntensity(value: Float) {
        rainIntensity = value.coerceIn(0f, 1f)
    }

    fun clear() {
        flashAge = 0f
        flashDuration = 0f
        cooldown = 60f
        bolt = null
        branches = emptyList()
    }

    private fun generateBolt() {
        val boltSide = 0.2f + Random.nextFloat() * 0.6f
        val segments = 7 + Random.nextInt(4)
        val main = mutableListOf(BoltPoint(boltSide, 0f))
        var x = boltSide
        val endY = 0.5f + Random.nextFloat() * 0.32f
        for (i in 1..segments) {
            val t = i / segments.toFloat()
            x += (Random.nextFloat() - 0.5f) * 0.11f
            x = x.coerceIn(0.05f, 0.95f)
            main.add(BoltPoint(x, t * endY))
        }
        bolt = main
        val branchList = mutableListOf<List<BoltPoint>>()
        val branchCount = 1 + Random.nextInt(2)
        for (b in 0 until branchCount) {
            val startIdx = 2 + Random.nextInt((main.size - 3).coerceAtLeast(1))
            val start = main[startIdx.coerceIn(0, main.lastIndex)]
            val branch = mutableListOf(start)
            var bx = start.x
            var by = start.y
            val dir = if (Random.nextBoolean()) 1f else -1f
            val steps = 2 + Random.nextInt(3)
            for (i in 0 until steps) {
                bx += dir * (0.03f + Random.nextFloat() * 0.06f)
                by += 0.04f + Random.nextFloat() * 0.07f
                bx = bx.coerceIn(0.03f, 0.97f)
                branch.add(BoltPoint(bx, by))
            }
            branchList.add(branch)
        }
        branches = branchList
    }

    fun tick(rainActive: Boolean, dt: Float = 1f) {
        if (flashDuration > 0f) {
            flashAge += dt
            if (flashAge >= flashDuration) {
                flashAge = 0f
                flashDuration = 0f
                bolt = null
                branches = emptyList()
                cooldown = 70f + Random.nextFloat() * 120f
            }
            return
        }
        if (!enabled || !rainActive || rainIntensity < 0.7f) return
        cooldown -= dt
        if (cooldown > 0f) return
        if (Random.nextFloat() > 0.018f + (rainIntensity - 0.7f) * 0.06f) return
        flashDuration = 4f + Random.nextFloat() * 6f
        flashAge = 0f
        cooldown = 80f + Random.nextFloat() * 140f
        bolt = null
        branches = emptyList()
        if (Random.nextFloat() < 0.55f) generateBolt()
    }

    private fun isFlashing() = flashDuration > 0f && flashAge < flashDuration

    private fun strokeBolt(canvas: Canvas, pts: List<BoltPoint>, w: Float, h: Float, width: Float, color: Int) {
        if (pts.size < 2) return
        strokePaint.color = color
        strokePaint.strokeWidth = width
        val path = Path()
        path.moveTo(pts[0].x * w, pts[0].y * h)
        for (i in 1 until pts.size) path.lineTo(pts[i].x * w, pts[i].y * h)
        canvas.drawPath(path, strokePaint)
    }

    fun draw(canvas: Canvas, clipPath: Path?, viewportW: Float, viewportH: Float) {
        if (!isFlashing()) return
        val t = flashAge / flashDuration
        val peak = if (t < 0.25f) t / 0.25f else 1f - (t - 0.25f) / 0.75f
        val alpha = (peak * 0.42f * 255).toInt()
        canvas.save()
        clipPath?.let { canvas.clipPath(it) }
        fillPaint.color = Color.argb((alpha * 0.5f).toInt(), 255, 250, 215)
        canvas.drawRect(0f, 0f, viewportW, viewportH, fillPaint)
        fillPaint.color = Color.argb((alpha * 0.3f).toInt(), 255, 235, 150)
        canvas.drawRect(0f, viewportH * 0.06f, viewportW, viewportH * 0.41f, fillPaint)
        val b = bolt
        if (b != null) {
            val boltFade = maxOf(0f, 1f - t * 1.4f)
            if (boltFade > 0.02f) {
                val scale = minOf(viewportW, viewportH) / 320f + 0.6f
                val a = (boltFade * 255).toInt()
                strokeBolt(canvas, b, viewportW, viewportH, 7f * scale, Color.argb(a, 10, 10, 20))
                for (br in branches) strokeBolt(canvas, br, viewportW, viewportH, 4.5f * scale, Color.argb(a, 10, 10, 20))
                strokeBolt(canvas, b, viewportW, viewportH, 4f * scale, Color.argb(a, 255, 214, 40))
                for (br in branches) strokeBolt(canvas, br, viewportW, viewportH, 2.4f * scale, Color.argb(a, 255, 214, 40))
            }
        }
        canvas.restore()
    }
}
